tools/gfx/cuda/render-cuda.cpp

```diff
@@ -0,0 +1,1506 @@
+        , programLayout(inProgramLayout)
+    {
+        for (UInt i =0; i< programLayout->getEntryPointCount(); i++)
+        {
+            entryPointLayouts.add(new CUDAShaderObjectLayout(
+                programLayout->getEntryPointByIndex(i)->getTypeLayout()));
+        }
+
+    }
+
+    int getKernelIndex(UnownedStringSlice kernelName)
+    {
+        for (int i = 0; i < (int)programLayout->getEntryPointCount(); i++)
+        {
+            auto entryPoint = programLayout->getEntryPointByIndex(i);
+            if (kernelName == entryPoint->getName())
+            {
+                return i;
+            }
+        }
+        return -1;
+    }
+
+    void getKernelThreadGroupSize(int kernelIndex, UInt* threadGroupSizes)
+    {
+        auto entryPoint = programLayout->getEntryPointByIndex(kernelIndex);
+        entryPoint->getComputeThreadGroupSize(3, threadGroupSizes);
+    }
+};
+
+class CUDAShaderObject : public ShaderObject
+{
+public:
+    RefPtr<MemoryCUDAResource> bufferResource;
+    RefPtr<CUDAShaderObjectLayout> layout;
+    List<RefPtr<CUDAShaderObject>> objects;
+    List<RefPtr<CUDAResourceView>> resources;
+
+    virtual SlangResult init(Renderer* renderer, CUDAShaderObjectLayout* typeLayout);
+
+    virtual slang::TypeLayoutReflection* getElementTypeLayout() override
+    {
+        return layout->typeLayout;
+    }
+    virtual Slang::Index getEntryPointCount() override { return 0; }
+    virtual ShaderObject* getEntryPoint(Slang::Index index) override { return nullptr; }
+    virtual SlangResult setData(ShaderOffset const& offset, void const* data, size_t size)
+    {
+        size = Math::Min(size, bufferResource->getDesc().sizeInBytes - offset.uniformOffset);
+        SLANG_CUDA_RETURN_ON_FAIL(cudaMemcpy(
+            (uint8_t*)bufferResource->m_cudaMemory + offset.uniformOffset,
+            data,
+            size,
+            cudaMemcpyHostToDevice));
+        return SLANG_OK;
+    }
+    virtual SlangResult getObject(ShaderOffset const& offset, ShaderObject** object)
+    {
+        auto subObjectIndex =
+            layout->m_bindingRanges[offset.bindingRangeIndex].baseIndex + offset.bindingArrayIndex;
+        if (subObjectIndex >= objects.getCount())
+        {
+            *object = nullptr;
+            return SLANG_OK;
+        }
+        *object = objects[subObjectIndex].Ptr();
+        return SLANG_OK;
+    }
+    virtual SlangResult setObject(ShaderOffset const& offset, ShaderObject* object)
+    {
+        auto subObjectIndex =
+            layout->m_bindingRanges[offset.bindingRangeIndex].baseIndex + offset.bindingArrayIndex;
+        SLANG_ASSERT(
+            offset.uniformOffset ==
+            layout->m_bindingRanges[offset.bindingRangeIndex].uniformOffset +
+                offset.bindingArrayIndex * sizeof(void*));
+        auto cudaObject = dynamic_cast<CUDAShaderObject*>(object);
+        if (subObjectIndex >= objects.getCount())
+            objects.setCount(subObjectIndex + 1);
```

Comment:
This step shouldn't be needed, right? We should be able to set the size of the `resources` array to `layout->m_resourceViewCount` at the time we create a `ShaderObject`, rather than try to dynamically resize stuff.

There's a certain amount of defensive code here trying to handle the case of the arrays being the wrong size, but it seems better to try and allocate the arrays at the exact *right* size at creation time.

(If we ever want to drill deep into performance stuff, we could switch to a model where the arrays of bindings on a `ShaderObject` are tail-allocated as part of the same memory as the object itself, so that we can use a single heap allocation for any object. In that case, the `ShaderObjectLayout` could pre-compute the byte size required for objects with that layout (including tail allocation), as well as byte offsets for each resource range.